.github/workflows/ci.yml

```diff
@@ -12,83 +12,60 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Check out Pulser
-        uses: actions/checkout@v2
-      - name: Set up Python
-        uses: actions/setup-python@v2
+        uses: actions/checkout@v3
+      - name: Pulser + flake8 setup
+        uses: ./.github/workflows/pulser-setup
         with:
-          python-version: 3.8
-      - name: Install Python dependencies
-        run: |
-          python -m pip install --upgrade pip
-          pip install -e ./pulser-core -e ./pulser-simulation
-          pip install -r requirements.txt
+          extra-packages: flake8
       - name: Lint with flake8
         run: flake8
   black:
     runs-on: ubuntu-latest
     steps:
       - name: Check out Pulser
-        uses: actions/checkout@v2
-      - name: Set up Python
-        uses: actions/setup-python@v2
+        uses: actions/checkout@v3
+      - name: Pulser + black setup
+        uses: ./.github/workflows/pulser-setup
         with:
-          python-version: 3.8
-      - name: Install black
-        run: |
-          python -m pip install --upgrade pip
-          pip install black
-          pip install 'black[jupyter]'
+          extra-packages: black
       - name: Check formatting with black
         run: black --check --diff .
   isort:
     runs-on: ubuntu-latest
     steps:
       - name: Check out Pulser
-        uses: actions/checkout@v2
-      - name: Set up Python
-        uses: actions/setup-python@v2
+        uses: actions/checkout@v3
+      - name: Pulser + isort setup
+        uses: ./.github/workflows/pulser-setup
         with:
-          python-version: 3.8
-      - name: Install Python dependencies
-        run: |
-          python -m pip install --upgrade pip
-          pip install -e ./pulser-core -e ./pulser-simulation
-          pip install -r requirements.txt
+          extra-packages: isort
       - name: Check import sorting with isort
         run: isort --check-only --diff .
   typing:
     runs-on: ubuntu-latest
     steps:
       - name: Check out Pulser
-        uses: actions/checkout@v2
-      - name: Set up Python
-        uses: actions/setup-python@v2
+        uses: actions/checkout@v3
+      - name: Pulser + mypy setup
+        uses: ./.github/workflows/pulser-setup
         with:
-          python-version: 3.8
-      - name: Install Python dependencies
-        run: |
-          python -m pip install --upgrade pip
-          pip install -e ./pulser-core -e ./pulser-simulation
-          pip install -r requirements.txt
+          extra-packages: '''mypy\|pytest'''
```

Comment:
nit: I find this syntax a big ugly but it's just a matter of taste.. I would prefer something a bit more classic (maybe this is very common) like `mypy,pytest` but it might be a pain to parse in the bash script then.

As a more general note (probably for another PR), It might be nice to add these dev dependencies in the `extras_require` section of the `setup.py` file instead of having an external `dev_requirements.txt`. You could separate optional dependencies in different categories, such as testing, linting and so on and instead of having this custom logic to identify packages to install (e.g. in this case pytest => find all packages which start with pytest in the `dev_requirements.txt`). Instead you could just do `pip install -e .[testing]` which would install all dependencies needed for testing.

The annoying thing with `extras_require` is that you have no way to install all `extras_require` categories except by listing them all, e.g. you can't `pip install -e .[all]` but instead `pip install -e .[linting, testing, blabla, blabla]`. To go the extra mile [this article](https://hanxiao.io/2019/11/07/A-Better-Practice-for-Managing-extras-require-Dependencies-in-Python/) suggests a cool idea to automatically generate the extras_require from a .txt file with a given format. The extras_require will then have an `all` category which lets you install everything with `pip install -e .[all]`

Anyway just a nice to have, nothing critical!

Reply:
I agree, for now it's just the regex that is given directly to grep, so there's no parsing required. I think that if this stops being an isolated case I would try to figure out how to parse it, but for now I think I'll stick with this little bit of ugliness. 

The `extras_require` is indeed a nice workaround, though admittedly out of scope for this PR perhaps. However, I will say I do like having the `dev_requirements` all neatly in a single file, rather than a bit more hidden away in the setup.py. Perhaps we can even find a way to combine the two, it would indeed be a nice to have!

Reply:
Yes definitely not a priority. If you prefer to have the dependencies isolated in a txt file then the article I linked might be a good hint for the future